File: 이준영/백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static int[] count;
	static int max_result= -1000000001;
	static int min_result= 1000000001;
	static int N=0;
	static int[] arr;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st= new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		arr = new int[N];
		count = new int[4];
 		st= new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) {
			arr[i]=Integer.parseInt(st.nextToken());
		}
		st= new StringTokenizer(br.readLine());
		for(int i=0;i<4;i++) {
			count[i]=Integer.parseInt(st.nextToken());
		}
		btk(0,arr[0]);
		System.out.println(max_result);
		System.out.println(min_result);
		
	}
	public static void btk(int cur,int result) {
		if (cur==N-1) {
			max_result=Math.max(max_result,result);
			min_result=Math.min(min_result,result);
			return;
		}
		for(int i=0;i<4;i++) {
			if(count[i]!=0) {
				count[i]-=1;
				switch (i) {
					case 0: 
						btk(cur+1,result+arr[cur+1]);
						break;
					case 1: 
						btk(cur+1,result-arr[cur+1]);
						break;
					case 2: 
						btk(cur+1,result*arr[cur+1]);
						break;
					case 3: 
						btk(cur+1,result/arr[cur+1]);
						break;
				}
				count[i]+=1;
			}
		}
		return;
	}
}

File: 이준영/백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
# 연산을 할 때 나눗셈만 주의하면 된다. (음수 나눗셈을 그냥 나눗셈을 하면 안됨)
def calc(num1, num2, c):            # c가 0,1,2,3일 때 각각 덧셈, 뺄셈, 곱셈, 나눗셈을 해서 결과를 리턴
    if c == 0:
        return num1 + num2
    if c == 1:
        return num1 - num2
    if c == 2:
        return num1 * num2

    if num1 >= 0:
        return num1 // num2
    else:
        return - ((-num1) // num2)


def btk(cur, res):                  # 연산을 몇 번 했는지, 결과
    global mn, mx

    if cur == N - 1:                # 연산 다하면 mx, mn을 갱신
        mx = max(mx, res)
        mn = min(mn, res)
        return

    for i in range(4):                          # 만약 해당 연산을 할 수 있으면 하자
        if cnt[i] >= 1:
            cnt[i] -= 1
            tmp = calc(res, arr[cur + 1], i)
            btk(cur + 1, tmp)
            cnt[i] += 1


N = int(input())
arr = list(map(int, input().split()))
cnt = list(map(int, input().split()))

mn = 10 ** 9 + 1
mx = - mn

btk(0, arr[0])

print(mx)
print(mn)
